x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

```diff
@@ -0,0 +1,155 @@
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamOutput;
+import org.elasticsearch.xpack.esql.plan.logical.join.Join;
+import org.elasticsearch.xpack.esql.plan.logical.join.JoinConfig;
+import org.elasticsearch.xpack.esql.plan.logical.join.JoinType;
+import org.elasticsearch.xpack.esql.plan.logical.local.LocalRelation;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.esql.expression.NamedExpressions.mergeOutputAttributes;
+
+/**
+ * Looks up values from the associated {@code tables}.
+ * The class is supposed to be substituted by a {@link Join}.
+ */
+public class Lookup extends UnaryPlan {
+    private final Expression tableName;
+    /**
+     * References to the input fields to match against the {@link #localRelation}.
+     */
+    private final List<NamedExpression> matchFields;
+    // initialized during the analysis phase for output and validation
+    // afterward, it is converted into a Join (BinaryPlan) hence why here it is not a child
+    private final LocalRelation localRelation;
+    private List<Attribute> lazyOutput;
+
+    public Lookup(
+        Source source,
+        LogicalPlan child,
+        Expression tableName,
+        List<NamedExpression> matchFields,
+        @Nullable LocalRelation localRelation
+    ) {
+        super(source, child);
+        this.tableName = tableName;
+        this.matchFields = matchFields;
+        this.localRelation = localRelation;
+    }
+
+    public Lookup(PlanStreamInput in) throws IOException {
+        super(Source.readFrom(in), in.readLogicalPlanNode());
+        this.tableName = in.readExpression();
+        this.matchFields = in.readCollectionAsList(i -> ((PlanStreamInput) i).readNamedExpression());
+        this.localRelation = in.readBoolean() ? new LocalRelation(in) : null;
+    }
+
+    public void writeTo(PlanStreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeLogicalPlanNode(child());
+        out.writeExpression(tableName);
+        out.writeCollection(matchFields, (o, v) -> ((PlanStreamOutput) o).writeNamedExpression(v));
+        if (localRelation == null) {
+            out.writeBoolean(false);
+        } else {
+            out.writeBoolean(true);
+            localRelation.writeTo(out);
+        }
+    }
+
+    public Expression tableName() {
+        return tableName;
+    }
+
+    public List<NamedExpression> matchFields() {
+        return matchFields;
+    }
+
+    public LocalRelation localRelation() {
+        return localRelation;
+    }
+
+    public JoinConfig joinConfig() {
+        List<Expression> conditions = new ArrayList<>(matchFields.size());
+        List<Attribute> rhsOutput = Join.makeReference(localRelation.output());
```

Comment:
This is indeed weird. Can't we just keep the field attributes as-is?

Reference attributes normally are needed when we create new attributes, but these are _not_ coming from fields - like in `| eval x = ...`, here `x = ...` is modeled via an `Alias` and calling `xAlias.toAttribute()` or `Expressions.asAttributes(xAlias)` will provide a corresponding `ReferenceAttribute`.

Reply:
Update: you made a nice comment in tests that explains why these need to be reference attributes, so never mind.